### PR TITLE
mon: preserve osd weight when marking osd out, then in

### PR DESCRIPTION
**Pull request by liewegas**
We already did this when the mon automatically marked an osd out and
then back in.  Do the same when the admin does it explicitly.

### CODE REVIEW

**Comment by tchaikov** [comment]
lgtm, needs rebase.

**Comment by yuriw** [comment]
jenkins test this please

**Comment by yuriw** [comment]
suspected this PR with @jdurgin for failures in http://pulpito.ceph.com/yuriw-2016-11-01_00:54:30-rados-wip-yuri-testing_2016_11_1-distro-basic-smithi/
so removing from the batch for now.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2016-11-05_20:56:53-rados-wip-kefu-testing---basic-mira/525011 fails.
```
/srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/osd/PG.cc: 5182: FAILED assert(get_min_upacting_features() & (1ULL<<51))

 ceph version 11.0.2-1173-g26682e9 (26682e950c5d9dfc1a536d09436015d01458f9f1)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x85) [0x55740ac6f6b5]
 2: (()+0x483dd9) [0x55740a6a1dd9]
 3: (PG::init(int, std::vector<int, std::allocator<int> > const&, int, std::vector<int, std::allocator<int> > const&, int, pg_history_t const&, std::map<unsigned int, pg_interval
_t, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, pg_interval_t> > >&, bool, ObjectStore::Transaction*)+0x515) [0x55740a6a6cf5]
 4: (OSD::_create_lock_pg(std::shared_ptr<OSDMap const>, spg_t, bool, bool, int, std::vector<int, std::allocator<int> >&, int, std::vector<int, std::allocator<int> >&, int, pg_hi
story_t, std::map<unsigned int, pg_interval_t, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, pg_interval_t> > >&, ObjectStore::Transaction&)+0x24d) [0x557
40a610d3d]
 5: (OSD::handle_pg_peering_evt(spg_t, pg_history_t const&, std::map<unsigned int, pg_interval_t, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, pg_interva
l_t> > >&, unsigned int, std::shared_ptr<PG::CephPeeringEvt>)+0x1506) [0x55740a62e856]
 6: (OSD::handle_pg_create(std::shared_ptr<OpRequest>)+0xe15) [0x55740a62fcd5]
 7: (OSD::dispatch_op(std::shared_ptr<OpRequest>)+0x220) [0x55740a6318b0]
 8: (OSD::_dispatch(Message*)+0x167) [0x55740a6320e7]
 9: (OSD::ms_dispatch(Message*)+0xa6) [0x55740a6327d6]
 10: (DispatchQueue::entry()+0x7ba) [0x55740ae17dfa]
 11: (DispatchQueue::DispatchThread::entry()+0xd) [0x55740acee2bd]
 12: (()+0x7dc5) [0x7fcbe7ef8dc5]
 13: (clone()+0x6d) [0x7fcbe6de021d]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
```

i am dropping this PR from my testing branch.

**Comment by tchaikov** [comment]
and http://pulpito.ceph.com/kchai-2016-11-06_09:02:10-rados-wip-kefu-testing2---basic-mira/, wip-kefu-testing2 is master + PR.

**Comment by liewegas** [comment]
Fixed, ready to retest

**Comment by liewegas** [comment]
passed tests (http://pulpito.ceph.com/sage-2016-11-08_02:52:28-rados-wip-sage-testing---basic-smithi/ .. most failures are bluestore related)

**Comment by liewegas** [comment]
@tchaikov this is fixed and passed my test, but could use another review since there are new patches/fixes included

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -2265,6 +2258,24 @@ bool OSDMonitor::prepare_boot(MonOpRequestRef op)
     else
       xi.features = m->get_connection()->get_features();
 
+    // mark in?
+    if ((g_conf->mon_osd_auto_mark_auto_out_in &&
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: The bug was that the original fix set new_xinfo without populating the entry with the previous map's xinfo structure, so ti clobbered other fields (in this case, features). In the fixed version we move the whole if down below where we have the xi already being properly populated and filled in with other info.

**Pull request merged**